### PR TITLE
[Enhancement] Support query single sync MV by using hint

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：
Single Sync MV cannot be queried for now,  it's hard to check what's wrong with it.

Fix:
- https://github.com/StarRocks/starrocks/issues/22401
- https://github.com/StarRocks/starrocks/issues/22352

Updates:
- Support to query single sync mv by using hint.
- Change MV's column to alias name for better query.(Note: only affects the new MVs).
- Add a config to control whether to support to rewrite Single Sync MV.

```
CREATE DATABASE test_materialized_view;
USE test_materialized_view;

CREATE TABLE `test_materialized_view_tbl1` (
  `k1` tinyint(4) NULL DEFAULT "0",
  `k2` varchar(64) NULL DEFAULT "",
  `k3` bigint NULL DEFAULT "0",
  `k4` varchar(64) NULL DEFAULT ""
) ENGINE=OLAP 
DUPLICATE KEY(`k1`, `k2`)
DISTRIBUTED BY HASH(`k1`, `k2`) BUCKETS 3
PROPERTIES (
"replication_num" = "1"
);
insert into test_materialized_view_tbl1 values (1, 'a', 1, 'aa'), (1, 'b', 2, 'bb'), (3, NULL, NULL, NULL);

CREATE MATERIALIZED VIEW test_materialized_view_tbl1_mv1 as select k1, k2, sum(k3) as sum_k3, count(k4) as count_k4 from test_materialized_view_tbl1 group by k1, k2;

select sleep(20);

select * from test_materialized_view_tbl1_mv1 [_SYNC_MV_] order by k1, k2;
select sum_k3 from test_materialized_view_tbl1_mv1 [_SYNC_MV_] order by k1, k2;
select count_k4, sum_k3 from test_materialized_view_tbl1_mv1 [_SYNC_MV_] order by k1, k2;
select k2, count_k4, sum_k3 from test_materialized_view_tbl1_mv1 [_SYNC_MV_] order by k1, k2;
select count(1) from test_materialized_view_tbl1_mv1 [_SYNC_MV_];
```

TODO: 
- Original Sync MV's column will not use alias as the MV table's column, I will refine this later.

eg:
```
CREATE MATERIALIZED VIEW test_materialized_view_tbl1_mv1 as select k1, k2, sum(k3) as sum_k3, count(k4) as count_k4 from test_materialized_view_tbl1 group by k1, k2;


mysql> select * from test_materialized_view_tbl1_mv1 [_SYNC_MV_] order by k1, k2;
+------+------+------+-------------+
| k1   | k2   | k3   | mv_count_k4 |
+------+------+------+-------------+
|    1 | a    |    1 |           1 |
|    1 | b    |    2 |           1 |
|    3 | NULL | NULL |           0 |
+------+------+------+-------------+
3 rows in set (0.01 sec)
```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 67 (38.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Database.java|0|12|00.00%|[645, 646, 647, 648, 649, 650, 651, 653, 654, 656, 658, 659]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|4|00.00%|[221, 222, 223, 224]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|4|00.00%|[2380, 2381, 2382, 2384]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[109]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1930, 1931]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|18|36|50.00%|[311, 312, 313, 314, 316, 317, 318, 319, 855, 856, 857, 858, 859, 860, 861, 862, 863, 865]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java [hunk 1]
```diff
@@ -835,7 +850,22 @@ private Table resolveTable(TableName tableName) {
...
-            Table table = metadataMgr.getTable(catalogName, dbName, tbName);
+            Table table = null;
+            if (tableRelation.isSyncMVQuery()) {
+                Pair<Table, MaterializedIndex> materializedIndex =
+                        metadataMgr.getMaterializedViewIndex(catalogName, dbName, tbName);
+                if (materializedIndex != null) {
+                    Table mvTable = materializedIndex.first;
+                    Preconditions.checkState(mvTable != null);
+                    Preconditions.checkState(mvTable instanceof OlapTable);
+                    OlapTable mvOlapTable = ((OlapTable) mvTable).copyOnlyForQuery();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java:861]: has concurrent issue. copyOnlyForQuery need lock protection.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 2]
```diff
@@ -301,8 +301,10 @@ private OptExpression logicalRuleRewrite(OptExpression tree, TaskContext rootTas
             }
         }
 
-        tree = new MaterializedViewRule().transform(tree, context).get(0);
-        deriveLogicalProperty(tree);
+        if (sessionVariable.isEnableSyncMaterializedViewRewrite()) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:304]: The new session variable should control we whether support directly query the sync mv, not effect the original sync mv rewrite.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 69 (37.68%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Database.java|0|12|00.00%|[645, 646, 647, 648, 649, 650, 651, 653, 654, 656, 658, 659]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|4|00.00%|[221, 222, 223, 224]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|4|00.00%|[2380, 2381, 2382, 2384]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[109]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|18|38|47.37%|[311, 312, 313, 314, 316, 317, 318, 319, 855, 856, 857, 858, 859, 860, 863, 864, 866, 867, 869, 872]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1930, 1931]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 2 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:304]: This is another thing. Add a comment, this config is only used to control whether to rewrite sync mv, so can be used in the test.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java]: fix it .

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 3]
```diff
@@ -987,6 +988,9 @@ public void setEnableParallelMerge(boolean enableParallelMerge) {
     @VarAttr(name = ENABLE_MATERIALIZED_VIEW_REWRITE)
     private boolean enableMaterializedViewRewrite = true;
 
+    @VarAttr(name = ENABLE_SYNC_MATERIALIZED_VIEW_REWRITE)
+    private boolean enableSyncMaterializedViewRewrite = true;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:992]: not a good name

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:992]: do you have a good idea? this config is only used to control whether to rewrite sync mv, so can be used in the test.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 4]
```diff
@@ -301,8 +301,11 @@ private OptExpression logicalRuleRewrite(OptExpression tree, TaskContext rootTas
...
+            // Add a config to decide whether to rewrite sync mv.
+            tree = new MaterializedViewRule().transform(tree, context).get(0);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:306]: MaterializedViewRule should check and respect the hint.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 44 / 87 (50.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Database.java|0|12|00.00%|[646, 647, 648, 649, 650, 651, 652, 654, 655, 657, 659, 660]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|4|00.00%|[221, 222, 223, 224]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|4|00.00%|[2380, 2381, 2382, 2384]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[109]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|18|38|47.37%|[311, 312, 313, 314, 316, 317, 318, 319, 855, 856, 857, 858, 859, 860, 863, 864, 866, 867, 869, 872]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1930, 1931]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRewriter.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MVProjectAggProjectScanRewrite.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/BestIndexRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneScanColumnRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 4 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:306]: fix it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 43 / 86 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Database.java|0|12|00.00%|[646, 647, 648, 649, 650, 651, 652, 654, 655, 657, 659, 660]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|4|00.00%|[221, 222, 223, 224]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|4|00.00%|[2380, 2381, 2382, 2384]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[109]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|18|38|47.37%|[311, 312, 313, 314, 316, 317, 318, 319, 855, 856, 857, 858, 859, 860, 863, 864, 866, 867, 869, 872]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1930, 1931]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRewriter.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MVProjectAggProjectScanRewrite.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/BestIndexRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneScanColumnRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|1|1|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 3 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:992]: ok, I misunderstood.

**Review by ABingHuang — APPROVED** [review]
LGTM

**Pull request merged**